File: python/shopping_cart.py
import math

from model_objects import ProductQuantity, SpecialOfferType, Discount


class ShoppingCart:

    def __init__(self):
        self._items = []
        self._product_quantities = {}

    @property
    def items(self):
        return self._items

    def add_item(self, product):
        self.add_item_quantity(product, 1.0)

    @property
    def product_quantities(self):
        return self._product_quantities

    def add_item_quantity(self, product, quantity):
        self._items.append(ProductQuantity(product, quantity))
        if product in self._product_quantities.keys():
            self._product_quantities[product] = self._product_quantities[product] + quantity
        else:
            self._product_quantities[product] = quantity

    def handle_offers(self, receipt, offers, catalog):
        for product in self._product_quantities.keys():
            quantity = self._product_quantities[product]
            if product in offers.keys():
                offer = offers[product]
                unit_price = catalog.unit_price(product)
                quantity_as_int = int(quantity)
                discount = None
                items_per_deal = 1
                if offer.offer_type == SpecialOfferType.THREE_FOR_TWO:
                    items_per_deal = 3

                elif offer.offer_type == SpecialOfferType.TWO_FOR_AMOUNT:
                    items_per_deal = 2
                    if quantity_as_int >= 2:
                        total = offer.argument * (quantity_as_int / items_per_deal) + quantity_as_int % 2 * unit_price
                        discount_amount = unit_price * quantity - total
                        discount = Discount(product, "2 for " + str(offer.argument), -discount_amount)

                if offer.offer_type == SpecialOfferType.FIVE_FOR_AMOUNT:
                    items_per_deal = 5

                number_of_complete_deals = math.floor(quantity_as_int / items_per_deal)
                if offer.offer_type == SpecialOfferType.THREE_FOR_TWO and quantity_as_int > 2:
                    discount_amount = quantity * unit_price - (
                                (number_of_complete_deals * 2 * unit_price) + quantity_as_int % 3 * unit_price)
                    discount = Discount(product, "3 for 2", -discount_amount)

                if offer.offer_type == SpecialOfferType.TEN_PERCENT_DISCOUNT:
                    discount = Discount(product, str(offer.argument) + "% off",
                                        -quantity * unit_price * offer.argument / 100.0)

                if offer.offer_type == SpecialOfferType.FIVE_FOR_AMOUNT and quantity_as_int >= 5:
                    total_discount_amount = unit_price * quantity - (
                                offer.argument * number_of_complete_deals + quantity_as_int % 5 * unit_price)
                    discount = Discount(product, str(items_per_deal) + " for " + str(offer.argument), -total_discount_amount)

                if discount:
                    receipt.add_discount(discount)

File: cpp/src/ShoppingCart.cpp
#include "ShoppingCart.h"

// Constants for magic numbers
namespace {
    constexpr double DEFAULT_ITEM_QUANTITY = 1.0;
    constexpr int TWO_FOR_AMOUNT_QUANTITY = 2;
    constexpr int THREE_FOR_TWO_QUANTITY = 3;
    constexpr int FIVE_FOR_AMOUNT_QUANTITY = 5;
    constexpr double PERCENTAGE_DIVISOR = 100.0;

    // Constants for discount description strings
    const std::string THREE_FOR_TWO_DESCRIPTION = "3 for 2";
    const std::string TWO_FOR_PREFIX = "2 for ";
    const std::string FOR_SEPARATOR = " for ";
    const std::string PERCENT_OFF_SUFFIX = "% off";
}

void addItemQuantity(const Product& product, double quantity);

std::vector<ProductQuantity> ShoppingCart::getItems() const {
    return items;
}

std::map<Product, double> ShoppingCart::getProductQuantities() const {
    return productQuantities;
}

void ShoppingCart::addItem(const Product& product) {
    addItemQuantity(product, DEFAULT_ITEM_QUANTITY);
}

void ShoppingCart::addItemQuantity(const Product& product, double quantity) {
    items.emplace_back(product, quantity);
    if (productQuantities.find(product) != productQuantities.end()) {
        productQuantities[product] += quantity;
    } else {
        productQuantities[product] = quantity;
    }
}

void ShoppingCart::handleOffers(Receipt& receipt, std::map<Product, Offer> offers, SupermarketCatalog* catalog) {
    for (const auto& productQuantity : productQuantities) {
        Product product = productQuantity.first;
        double quantity = productQuantity.second;
        if (offers.find(product) != offers.end()) {
            auto offer = offers[product];
            double unitPrice = catalog->getUnitPrice(product);
            int quantityAsInt = (int) quantity;
            Discount* discount = nullptr;
            int x = 1;

            if (offer.getOfferType() == SpecialOfferType::ThreeForTwo) {
                x = THREE_FOR_TWO_QUANTITY;
            } else if (offer.getOfferType() == SpecialOfferType::TwoForAmount) {
                x = TWO_FOR_AMOUNT_QUANTITY;
                if (quantityAsInt >= TWO_FOR_AMOUNT_QUANTITY) {
                    double total = offer.getArgument() * (quantityAsInt / x) + quantityAsInt % TWO_FOR_AMOUNT_QUANTITY * unitPrice;
                    double discountN = unitPrice * quantity - total;
                    discount = new Discount(TWO_FOR_PREFIX + std::to_string(offer.getArgument()), -discountN, product);
                }
            } if (offer.getOfferType() == SpecialOfferType::FiveForAmount) {
                x = FIVE_FOR_AMOUNT_QUANTITY;
            }
            int numberOfXs = quantityAsInt / x;
            if (offer.getOfferType() == SpecialOfferType::ThreeForTwo && quantityAsInt > TWO_FOR_AMOUNT_QUANTITY) {
                double discountAmount = quantity * unitPrice - ((numberOfXs * TWO_FOR_AMOUNT_QUANTITY * unitPrice) + quantityAsInt % THREE_FOR_TWO_QUANTITY * unitPrice);
                discount = new Discount(THREE_FOR_TWO_DESCRIPTION, -discountAmount, product);
            }
            if (offer.getOfferType() == SpecialOfferType::TenPercentDiscount) {
                discount = new Discount(std::to_string(offer.getArgument()) + PERCENT_OFF_SUFFIX, -quantity * unitPrice * offer.getArgument() / PERCENTAGE_DIVISOR, product);
            }
            if (offer.getOfferType() == SpecialOfferType::FiveForAmount && quantityAsInt >= FIVE_FOR_AMOUNT_QUANTITY) {
                double discountTotal = unitPrice * quantity - (offer.getArgument() * numberOfXs + quantityAsInt % FIVE_FOR_AMOUNT_QUANTITY * unitPrice);
                discount = new Discount(std::to_string(x) + FOR_SEPARATOR + std::to_string(offer.getArgument()), -discountTotal, product);
            }
            if (discount != nullptr)
                receipt.addDiscount(*discount);
        }
    }
}